File: Packages/TriathlonEngine/Sources/TriathlonEngine/Analysis/PersonalRecords.swift
import Foundation

public struct PersonalRecord: Sendable, Equatable, Identifiable {
    public var id: String { sportKey + label }
    public var sportKey: String
    public var label: String
    public var value: String
    public var date: Date
    public init(sportKey: String, label: String, value: String, date: Date) {
        self.sportKey = sportKey; self.label = label; self.value = value; self.date = date
    }
}

/// Calcule les records personnels à partir des activités réalisées.
public struct PersonalRecords: Sendable {
    public init() {}

    public func compute(from activities: [CompletedActivity]) -> [PersonalRecord] {
        var records: [PersonalRecord] = []

        for sport in [Sport.swim, .bike, .run] {
            let acts = activities.filter { $0.sport == sport }
            guard !acts.isEmpty else { continue }

            // Plus longue distance.
            if let longest = acts.filter({ $0.distanceM != nil }).max(by: { ($0.distanceM ?? 0) < ($1.distanceM ?? 0) }),
               let d = longest.distanceM {
                records.append(.init(sportKey: sport.rawValue, label: "Plus longue distance",
                                     value: km(d), date: longest.start))
            }
            // Plus longue durée.
            if let longest = acts.max(by: { $0.duration < $1.duration }) {
                records.append(.init(sportKey: sport.rawValue, label: "Plus longue durée",
                                     value: duration(longest.duration), date: longest.start))
            }
            // Records spécifiques.
            switch sport {
            case .bike:
                if let best = acts.compactMap({ a in (a.normalizedPowerW ?? a.avgPowerW).map { (a, $0) } })
                    .max(by: { $0.1 < $1.1 }) {
                    records.append(.init(sportKey: sport.rawValue, label: "Meilleure puissance",
                                         value: "\(best.1) W", date: best.0.start))
                }
            case .run:
                if let best = acts.filter({ ($0.distanceM ?? 0) >= 3000 })
                    .compactMap({ a in a.avgPaceSecPerKm.map { (a, $0) } })
                    .min(by: { $0.1 < $1.1 }) {
                    records.append(.init(sportKey: sport.rawValue, label: "Meilleure allure (≥3 km)",
                                         value: pace(best.1), date: best.0.start))
                }
            case .swim:
                if let best = acts.filter({ ($0.distanceM ?? 0) > 0 })
                    .map({ a in (a, 100.0 / (a.distanceM! / a.duration)) })
                    .min(by: { $0.1 < $1.1 }) {
                    records.append(.init(sportKey: sport.rawValue, label: "Meilleure allure /100 m",
                                         value: pace100(best.1), date: best.0.start))
                }
            default: break
            }
        }
        return records
    }

    private func km(_ m: Double) -> String { String(format: "%.1f km", m / 1000) }
    private func duration(_ s: TimeInterval) -> String {
        let t = Int(s.rounded()); let h = t / 3600, m = (t % 3600) / 60
        return h > 0 ? "\(h) h \(String(format: "%02d", m))" : "\(m) min"
    }
    private func pace(_ secPerKm: Double) -> String {
        let s = Int(secPerKm.rounded()); return String(format: "%d:%02d/km", s / 60, s % 60)
    }
    private func pace100(_ secPer100: Double) -> String {
        let s = Int(secPer100.rounded()); return String(format: "%d:%02d/100m", s / 60, s % 60)
    }
}

File: web/src/engine/analysis/personalRecords.ts
// Portage de Analysis/PersonalRecords.swift.
import type { CompletedActivity } from '../domain/activity';
import type { Sport } from '../domain/types';

export interface PersonalRecord {
  sportKey: string;
  label: string;
  value: string;
  date: Date;
}

function km(m: number): string {
  return `${(m / 1000).toFixed(1)} km`;
}
function durationText(s: number): string {
  const t = Math.round(s);
  const h = Math.trunc(t / 3600);
  const m = Math.trunc((t % 3600) / 60);
  return h > 0 ? `${h} h ${String(m).padStart(2, '0')}` : `${m} min`;
}
function paceText(secPerKm: number): string {
  const s = Math.round(secPerKm);
  return `${Math.trunc(s / 60)}:${String(s % 60).padStart(2, '0')}/km`;
}
function pace100Text(secPer100: number): string {
  const s = Math.round(secPer100);
  return `${Math.trunc(s / 60)}:${String(s % 60).padStart(2, '0')}/100m`;
}

export class PersonalRecords {
  compute(activities: CompletedActivity[]): PersonalRecord[] {
    const records: PersonalRecord[] = [];
    const sports: Sport[] = ['swim', 'bike', 'run'];

    for (const sport of sports) {
      const acts = activities.filter((a) => a.sport === sport);
      if (acts.length === 0) continue;

      const withDist = acts.filter((a) => a.distanceM != null);
      if (withDist.length > 0) {
        const longest = withDist.reduce((best, a) => ((a.distanceM ?? 0) > (best.distanceM ?? 0) ? a : best));
        records.push({ sportKey: sport, label: 'Plus longue distance', value: km(longest.distanceM!), date: longest.start });
      }

      const longestDur = acts.reduce((best, a) => (a.duration > best.duration ? a : best));
      records.push({ sportKey: sport, label: 'Plus longue durée', value: durationText(longestDur.duration), date: longestDur.start });

      if (sport === 'bike') {
        const withPower = acts
          .map((a) => ({ a, p: a.normalizedPowerW ?? a.avgPowerW }))
          .filter((x): x is { a: CompletedActivity; p: number } => x.p != null);
        if (withPower.length > 0) {
          const best = withPower.reduce((b, x) => (x.p > b.p ? x : b));
          records.push({ sportKey: sport, label: 'Meilleure puissance', value: `${best.p} W`, date: best.a.start });
        }
      } else if (sport === 'run') {
        const eligible = acts
          .filter((a) => (a.distanceM ?? 0) >= 3000)
          .map((a) => ({ a, pace: a.avgPaceSecPerKm }))
          .filter((x): x is { a: CompletedActivity; pace: number } => x.pace != null);
        if (eligible.length > 0) {
          const best = eligible.reduce((b, x) => (x.pace < b.pace ? x : b));
          records.push({ sportKey: sport, label: 'Meilleure allure (≥3 km)', value: paceText(best.pace), date: best.a.start });
        }
      } else if (sport === 'swim') {
        const eligible = acts
          .filter((a) => (a.distanceM ?? 0) > 0)
          .map((a) => ({ a, pace: 100.0 / (a.distanceM! / a.duration) }));
        if (eligible.length > 0) {
          const best = eligible.reduce((b, x) => (x.pace < b.pace ? x : b));
          records.push({ sportKey: sport, label: 'Meilleure allure /100 m', value: pace100Text(best.pace), date: best.a.start });
        }
      }
    }
    return records;
  }
}
